src/interface_vault_operator_peers.py

```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/env python3
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Work in progress interface for Vault peer relations."""
+
+import logging
+
+from ops.framework import EventBase, EventSource, Object, ObjectEvents
+from ops.model import Relation
+
+
+class HasPeersEvent(EventBase):
+    """Has peers event."""
+
+    pass
+
+
+class ReadyPeersEvent(EventBase):
+    """Ready Peers event."""
+
+    pass
+
+
+class VaultOperatorPeerEvents(ObjectEvents):
+    """Vault Operator Peer events."""
+
+    has_peers = EventSource(HasPeersEvent)
+    ready_peers = EventSource(ReadyPeersEvent)
+
+
+class VaultOperatorPeers(Object):
+    """Handles logic between instances of the vault operator."""
+
+    ROOT_TOKEN = "root-token"
+    ROOT_CA = "root-ca"
+    UNSEAL_KEY = "unseal-key"
+
+    on = VaultOperatorPeerEvents()
+
+    def __init__(self, charm, relation_name):
+        super().__init__(charm, relation_name)
+        self.relation_name = relation_name
+        self.framework.observe(charm.on[relation_name].relation_joined, self.on_joined)
+        self.framework.observe(charm.on[relation_name].relation_changed, self.on_changed)
+
+    @property
+    def peers_rel(self) -> Relation:
+        """Returns peer relation."""
+        return self.framework.model.get_relation(self.relation_name)  # type: ignore[return-value]
+
+    def on_joined(self, event):
+        """On peer relation joined."""
+        logging.info("VaultOperatorPeers on_joined")
+        self.on.has_peers.emit()
+
+    def on_changed(self, event):
+        """On peer relation changed."""
+        logging.info("VaultOperatorPeers on_changed")
+        self.on.ready_peers.emit()
+
+    def set_root_token(self, token: str):
+        """Sets root token in relation data."""
+        logging.info("Setting root token")
+        self.peers_rel.data[self.peers_rel.app][self.ROOT_TOKEN] = token  # type: ignore[index]
+
+    def set_root_ca(self, ca: str) -> None:
+        """Sets root ca in peer relation data."""
+        if self.root_ca is not None:
+            return
```

Comment:
Feels like this warrants a log message to explain why we bailed?

Reply:
Agreed